src/client/src/producer.rs

```diff
@@ -62,21 +78,57 @@ impl TopicProducer {
         V: Into<Vec<u8>>,
         I: IntoIterator<Item = (Option<K>, V)>,
     {
-        let replica = ReplicaKey::new(&self.topic, 0);
-        let spu_client = self.pool.create_serial_socket(&replica).await?;
-        debug!(
-            addr = spu_client.config().addr(),
-            "Connected to replica leader:"
-        );
-        let records: Vec<_> = records
+        let topics = self.pool.metadata.topics();
+        let topic_spec = topics
+            .lookup_by_key(&self.topic)
+            .await?
+            .ok_or_else(|| FluvioError::TopicNotFound(self.topic.to_string()))?
+            .spec;
+        let partition_count = topic_spec.partitions();
+        let partition_config = PartitionerConfig { partition_count };
+
+        let entries = records
             .into_iter()
-            .map(|(key, value): (Option<K>, V)| {
-                let key = key.map(|k| Bytes::from(k.into()));
-                let value = Bytes::from(value.into());
-                (key, value)
-            })
-            .collect();
-        send_records_raw(spu_client, &replica, records).await?;
+            .map::<(Option<Vec<u8>>, Vec<u8>), _>(|(k, v)| (k.map(|it| it.into()), v.into()))
+            .map(|(key, value)| {
+                let key = key.map(|it| DefaultAsyncBuffer::new(it));
+                let value = DefaultAsyncBuffer::new(value);
+                DefaultRecord::from((key, value))
+            });
+
+        // Calculate the partition for each entry
+        // Use a block scope to ensure we drop the partitioner lock
+        let records_by_partition = {
+            let mut iter = Vec::new();
+            let mut partitioner = self.partitioner.lock().await;
+            partitioner.update_config(partition_config);
+
+            for record in entries {
```

Comment:
I tried but I actually cannot. The body of this loop uses the `partitioner`, so if that body is moved into a closure, it then borrows the `partitioner` for longer than allowed (because iterators are lazy and using `.map` simply saves the closure object)

Reply:
k. fair enough